flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/FailoverRegion.java

```diff
@@ -152,10 +160,16 @@ private void cancel(final long globalModVersionOfFailover) {
 					}
 
 					final FutureUtils.ConjunctFuture<Void> allTerminal = FutureUtils.waitForAll(futures);
-					allTerminal.thenAcceptAsync(
-						(Void value) -> allVerticesInTerminalState(globalModVersionOfFailover),
-						executor);
-
+					final Consumer<Void> recoveryAction =
+						(Void value) -> allVerticesInTerminalState(globalModVersionOfFailover);
+
+					if (testExecutor == null) {
+						// branch for production code.
+						allTerminal.thenAccept(recoveryAction);
+					} else {
+						// this branch should only be taken for some unit tests.
+						allTerminal.thenAcceptAsync(recoveryAction, testExecutor);
```

Comment:
It is the `PipelinedRegionFailoverConcurrencyTest` because we need to pass in a custom executor with manual triggering to step through execution in a particular order.

Reply:
Alternatively, we could also generally pass in the excutor, in the production code the main thread executor. However, this would require to change at what time the failover strategy is created, because right now, the executor is passed in in `start(...)` which happens after the failover stratey is created.